client/network/src/service/tests.rs

```diff
@@ -0,0 +1,262 @@
+			}
+			if rand::random::<u8>() % 5 >= 3 {
+				node2.write_notification(node1.local_peer_id().clone(), ENGINE_ID, b"hello world".to_vec());
+			}
+
+			// Also randomly disconnect the two nodes from time to time.
+			if rand::random::<u8>() % 20 == 0 {
+				node1.disconnect_peer(node2.local_peer_id().clone());
+			}
+			if rand::random::<u8>() % 20 == 0 {
+				node2.disconnect_peer(node1.local_peer_id().clone());
+			}
+
+			// Grab next event from either `events_stream1` or `events_stream2`.
+			let next_event = {
+				let next1 = events_stream1.next();
+				let next2 = events_stream2.next();
+				// We also await on a small timer, otherwise it is possible for the test to wait
+				// forever while nothing at all happens on the network.
+				let continue_test = futures_timer::Delay::new(Duration::from_millis(20));
+				match future::select(future::select(next1, next2), continue_test).await {
+					future::Either::Left((future::Either::Left((Some(ev), _)), _)) =>
+						future::Either::Left(ev),
+					future::Either::Left((future::Either::Right((Some(ev), _)), _)) =>
+						future::Either::Right(ev),
+					future::Either::Right(_) => continue,
+					_ => break,
+				}
+			};
+
+			match next_event {
+				future::Either::Left(Event::NotificationStreamOpened { remote, engine_id, .. }) => {
+					assert!(!node1_to_node2_open);
+					node1_to_node2_open = true;
+					assert_eq!(remote, *node2.local_peer_id());
+					assert_eq!(engine_id, ENGINE_ID);
+				}
+				future::Either::Right(Event::NotificationStreamOpened { remote, engine_id, .. }) => {
+					assert!(!node2_to_node1_open);
+					node2_to_node1_open = true;
+					assert_eq!(remote, *node1.local_peer_id());
+					assert_eq!(engine_id, ENGINE_ID);
+				}
+				future::Either::Left(Event::NotificationStreamClosed { remote, engine_id, .. }) => {
+					assert!(node1_to_node2_open);
+					node1_to_node2_open = false;
+					assert_eq!(remote, *node2.local_peer_id());
+					assert_eq!(engine_id, ENGINE_ID);
+				}
+				future::Either::Right(Event::NotificationStreamClosed { remote, engine_id, .. }) => {
+					assert!(node2_to_node1_open);
+					node2_to_node1_open = false;
+					assert_eq!(remote, *node1.local_peer_id());
+					assert_eq!(engine_id, ENGINE_ID);
+				}
+				future::Either::Left(Event::NotificationsReceived { remote, .. }) => {
+					assert!(node1_to_node2_open);
+					assert_eq!(remote, *node2.local_peer_id());
+					if rand::random::<u8>() % 5 >= 4 {
+						node1.write_notification(
+							node2.local_peer_id().clone(),
+							ENGINE_ID,
+							b"hello world".to_vec()
+						);
+					}
+				}
+				future::Either::Right(Event::NotificationsReceived { remote, .. }) => {
+					assert!(node2_to_node1_open);
+					assert_eq!(remote, *node1.local_peer_id());
+					if rand::random::<u8>() % 5 >= 4 {
+						node2.write_notification(
+							node1.local_peer_id().clone(),
+							ENGINE_ID,
+							b"hello world".to_vec()
+						);
+					}
+				}
+				_ => {}
```

Comment:
```suggestion
				future::Either::Left(Event::Dht { .. }) => {}
				future::Either::Right(Event::Dht { .. }) => {}
```

Reply:
We could add other non-DHT events later.

Reply:
Yes, this is why I listed the the current ones so that when adding new ones the implications for this test would be checked.